lib/rbac/approval_access.rb

```diff
@@ -0,0 +1,133 @@
+module RBAC
+  class ApprovalAccess < Access
+    include RBAC::Permissions
+
+    ADMIN_ROLE = 'Approval Administrator'.freeze
+    APPROVER_ROLE = 'Approval Approver'.freeze
+    APP_NAME = 'approval'.freeze
+
+    def initialize(resource, verb)
+      super(resource, verb)
+      @admin = RBAC::Roles.assigned_role?(ADMIN_ROLE)
+      @approver = RBAC::Roles.assigned_role?(APPROVER_ROLE)
+      self.process
+    end
+
+    # permission level check
+    def resource_accessible?
+      admin? || acl.any? || owner_acls.any?
+    end
+
+    # instance level check
+    def resource_instance_accessible?(resource, resource_id)
+      admin? || approvable?(resource, resource_id) || owned?(resource, resource_id)
+    end
+
+    def admin?
+      @admin
+    end
+
+    def approver?
+      @approver
+    end
+
+    # current RBAC user
+    def self.whoami?
+      RBAC::Service.call(RBACApiClient::AccessApi) do |api|
+        api.api_client.config.username
+      end
+    end
+
+    # check if approver can process the #{resource} with #{id}
+    def approvable?(resource, id)
+      approver? && approver_id_list(resource)&.include?(id.to_i)
+    end
+
+    # check if regular requester own the #{resource} with #{id}
+    def owned?(resource, id)
+      owner_id_list(resource)&.include?(id.to_i)
+    end
+
+    # resource ids approver can approve
+    def approver_id_list(resource)
+      id_list(resource, approver_request_ids)
+    end
+
+    # resource ids owner owns
+    def owner_id_list(resource)
+      id_list(resource, owner_request_ids)
+    end
+
+    private
+
+    # Owner access list for the #{resource} and action #{verb}
+    def owner_acls
+      regexp = Regexp.new(":(#{@resource}|\\*):(#{@verb}|\\*)")
+      requester_acls.select do |item|
+        regexp.match(item.permission)
+      end
+    end
+
+    def approver_acls
+      approval_access = RBAC::Access.new('workflows', 'approve')
+      approval_access.process
+      approval_access.acl
+    end
+
+    # Request ids owned by requester
+    def owner_request_ids
+      Request.by_owner.pluck(:id).sort
+    end
+
+    # Request ids approver can process
+    def approver_request_ids
+      Request.where(:workflow_id => workflow_ids).pluck(:id).sort
```

Comment:
@hsong-rh All ActiveRecord access should be in your mixin and not in the lib/....